serving/docker/partition/run_multi_node_setup.py

```diff
@@ -0,0 +1,267 @@
+    remove_option_from_properties,
+    update_kwargs_with_env_vars,
+)
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class MultiNodeSetupHandler:
+
+    def __init__(self):
+        leader_hostname = os.environ.get("DJL_LEADER_ADDR", "")
+        if not leader_hostname:
+            raise ValueError(
+                f"Hostname for DJL_LEADER_ADDR unavailable. Cannot obtain model info. Multi-node setup will not proceed."
+            )
+
+        self.cluster_leader_hostname = leader_hostname
+        self.cluster_management_port = 8888
+
+        self.endpoint_url = f"http://{self.cluster_leader_hostname}:{self.cluster_management_port}/cluster/"
+
+    def hash(self, model_id):
+        try:
+            sha256 = hashlib.sha256()
+            sha256.update(model_id.encode("utf-8"))
+            digest = sha256.digest()
+            hex_digest = digest[:20].hex()
+            return hex_digest
+        except Exception as e:
+            raise AssertionError(
+                "Error occurred while hashing the input model_url: ", e)
+
+    def get_cache_dir(self):
+        cache_dir = os.environ.get("DJL_CACHE_DIR")
+        if not cache_dir:
+            home_dir = Path(os.path.expanduser("~"))
+            if not os.access(home_dir, os.W_OK):
+                cache_dir = Path(os.environ.get("TEMP", "/tmp"))
+            else:
+                cache_dir = home_dir / ".djl.ai"
+        else:
+            cache_dir = Path(cache_dir)
+
+        return cache_dir
+
+    @retry(stop_max_delay=60000, wait_fixed=2000)
+    def get_model_info_and_download_model(self):
+        """
+        Queries the Leader Address in LWS to obtain model_name, and model_url.
+        Downloads the model to cache
+        """
+        try:
+            logger.info(
+                f"Getting model info from leader node to download model")
+
+            endpoint_url = self.endpoint_url + "models"
+            response = requests.get(endpoint_url)
+
+            if response.status_code == 200:
+                data = response.json()
+                if type(data) != dict:
+                    raise ValueError(
+                        f"Expected a json dictionary response but received data of type: {type(data)}"
+                    )
+
+                for model_id, model_url in data.items():
+                    logger.info(f"Model Name: {model_id}")
+                    logger.info(f"Model URL: {model_url}")
+
+                    if not model_url.startswith("s3://"):
+                        # A path to serving.properties is returned
+                        properties = load_properties(model_url)
+                        if "option.model_id" in properties:
+                            model_url = properties["option.model_id"]
+                        elif "OPTION_MODEL_ID" in os.environ:
+                            model_url = os.environ.get("OPTION_MODEL_ID")
+                    if not model_url.endswith("/"):
```

Comment:
check if model_url is None or start with s3 here, and move append "/" logic into download function